dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/tasks/sub_process.py

```diff
@@ -0,0 +1,65 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Task sub_process."""
+
+from typing import Dict
+
+from pydolphinscheduler.constants import TaskType
+from pydolphinscheduler.core.process_definition import ProcessDefinitionContext
+from pydolphinscheduler.core.task import Task, TaskParams
+from pydolphinscheduler.java_gateway import launch_gateway
+
+
+class SubProcessTaskParams(TaskParams):
+    """Parameter only for Sub Process task type."""
+
+    def __init__(self, process_definition_code, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.process_definition_code = process_definition_code
+
+
+class SubProcess(Task):
+    """Task SubProcess object, declare behavior for SubProcess task to dolphinscheduler."""
+
+    def __init__(self, name: str, process_definition_name: str, *args, **kwargs):
+        self._process_definition_name = process_definition_name
+        self._process_definition = {}
+        task_params = SubProcessTaskParams(
+            process_definition_code=self.get_process_definition_code(),
+        )
+        super().__init__(name, TaskType.SUB_PROCESS, task_params, *args, **kwargs)
+
+    def get_process_definition_code(self) -> str:
+        """Get process definition code, a wrapper for :func:`get_process_definition_info`."""
+        return self.get_process_definition_info(self._process_definition_name).get(
+            "code"
+        )
+
+    def get_process_definition_info(self, process_definition_name: str) -> Dict:
+        """Get process definition info from java gateway, contains process definition id, name, code."""
+        if self._process_definition:
+            return self._process_definition
+        else:
+            gateway = launch_gateway()
+            process_definition = ProcessDefinitionContext.get()
```

Comment:
Hi @devosend, we already have attribute `process_definition` in base `Task` class in 
https://github.com/apache/dolphinscheduler/blob/d93248acbc3f7f1db299f7573b1403575f6da2a0/dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/core/task.py#L152-L154
So maybe you could just use statement `self.process_definition` to call it.

But take of one edge case, when user assign outside the `ProcessDefinition` context and not assign `process_definiton` by them self, we would got null as return. So maybe we should check `self.process_definiton` value before we use it. if value is `None` we should raise error. I recommend you add new exception in https://github.com/apache/dolphinscheduler/blob/d93248acbc3f7f1db299f7573b1403575f6da2a0/dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/exceptions.py#L18 named `PyDSProcessDefinitionNotAssignException` or something like that.

Reply:
Hi @zhongjiajie , thank's for your advice.
Because `process_definition` needs parent class initialization, which will lead to parent-child class initialization depending on each other, I think `process_definition` needs to be acquired in subprocess and then passed to parent class initialization. 
like this:
```
        self.process_definition = ProcessDefinitionContext.get()
        task_params = SubProcessTaskParams(
            process_definition_code=self.get_process_definition_code(),
        )
        super().__init__(
            name,
            TaskType.SUB_PROCESS,
            task_params,
            process_definition=self.process_definition,
            *args,
            **kwargs
        )
```
 What do you think?

Reply:
Hi @devosend , I do no think it will, for now we have two way to use task in workflow-as-code, with context or direct assignment to task attribute.

```py
# Context mode
# In this case, sub_process would get `self.process_definiton`
# from parent `Task`, and `Task` `self.process_definiton`
# would get from `ProcessDefinitionContext`
with ProcessDefinition(...) as pd:
   sub_process = SubProcess(name="test", process_name="sub-name")

# Seperate mode
# We would directly assign ProcessDefinition instance to
# `SubProcess` parameter `process_definiton` and we could use it directly.
pd = ProcessDefinition(...)
sub_process = SubProcess(name="test", process_name="sub-name", process_definiton=pd)
```

Reply:
Did I miss something in this case?